storm-core/src/clj/org/apache/storm/daemon/executor.clj

```diff
@@ -508,7 +509,7 @@
         ^Integer max-spout-pending (if max-spout-pending (int max-spout-pending))        
         last-active (atom false)        
         spouts (ArrayList. (map :object (vals task-datas)))
-        rand (Random. (Utils/secureRandomLong))
+        rand (XORShiftRandom. (Utils/secureRandomLong))
```

Comment:
We cannot use a thread local random here, because it is placed into functions that are not called from a single thread.  What is more the implementation of XORShiftRandom is that if we do lose the race, it is likely that we will repeat a number which could result in tuples not being tracked properly.